framework/finalizer.go

```diff
@@ -0,0 +1,99 @@
+package framework
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	finalizerName = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string   `json:"op"`
+	Path  string   `json:"path"`
+	Value []string `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, err
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return true, nil // object has been marked for deletion, we should ignore it.
+	}
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return false, nil // object already has the finalizer.
+	}
+
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return false, err
+	}
+	patch := fmt.Sprintf(`{"op":"add","value":%q,"path":"/metadata/finalizers/1"}`, finalizerName)
+	res := c.Patch(types.JSONPatchType).
+		AbsPath(accessor.GetSelfLink()).
+		Body(patch).
+		Do()
+	if res.Error() != nil {
+		return false, res.Error()
+	}
+
+	return true, nil
+}
+
+func (f *Framework) removeFinalizer(obj interface{}) error {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	if !containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return nil // resource has no finalizer, probably migration.
+	}
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return err
+	}
+
+	patch := patchSpec{
+		Op:    "replace",
+		Value: removeFinalizer(accessor.GetFinalizers(), finalizerName),
+		Path:  "/metadata/finalizers",
+	}
+	p, err := json.Marshal(patch)
+	if err != nil {
+		return err
+	}
+	res := c.Patch(types.JSONPatchType).
+		AbsPath(accessor.GetSelfLink()).
+		Body(p).
+		Do()
```

Comment:
I think this wrapping does more harm than good for readability.

Reply:
I saw it like this in the k8s api, I am fine with either so I changed it to a single line for now.